File: Workpiece/Summoner/LocalPackages/com.gamelovers.mcp-unity/Editor/Tools/TransformTools.cs
using UnityEngine;
using UnityEditor;
using McpUnity.Unity;
using Newtonsoft.Json.Linq;

namespace McpUnity.Tools
{
    /// <summary>
    /// Tool for moving a GameObject's position in the Unity Editor.
    /// Supports world/local space and absolute/relative positioning.
    /// </summary>
    public class MoveGameObjectTool : McpToolBase
    {
        public MoveGameObjectTool()
        {
            Name = "move_gameobject";
            Description = "Moves a GameObject to a new position. Supports world/local space and absolute/relative positioning.";
            IsAsync = false;
        }

        public override JObject Execute(JObject parameters)
        {
            // Find the GameObject
            var findResult = TransformToolUtils.FindGameObject(parameters);
            if (findResult.Error != null)
                return findResult.Error;

            GameObject gameObject = findResult.GameObject;
            Transform transform = gameObject.transform;

            // Extract position
            JObject positionObj = parameters["position"] as JObject;
            if (positionObj == null)
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    "Required parameter 'position' not provided",
                    "validation_error"
                );
            }

            Vector3 position = new Vector3(
                positionObj["x"]?.ToObject<float>() ?? 0f,
                positionObj["y"]?.ToObject<float>() ?? 0f,
                positionObj["z"]?.ToObject<float>() ?? 0f
            );

            // Get space and relative flags
            string space = parameters["space"]?.ToObject<string>() ?? "world";
            bool relative = parameters["relative"]?.ToObject<bool>() ?? false;

            // Record undo
            Undo.RecordObject(transform, "Move GameObject");

            // Apply the position change
            if (space.ToLower() == "local")
            {
                if (relative)
                    transform.localPosition += position;
                else
                    transform.localPosition = position;
            }
            else // world space
            {
                if (relative)
                    transform.position += position;
                else
                    transform.position = position;
            }

            EditorUtility.SetDirty(gameObject);

            // Return result with new position
            return new JObject
            {
                ["success"] = true,
                ["type"] = "text",
                ["message"] = $"GameObject '{gameObject.name}' moved successfully.",
                ["instanceId"] = gameObject.GetInstanceID(),
                ["name"] = gameObject.name,
                ["path"] = TransformToolUtils.GetGameObjectPath(gameObject),
                ["position"] = new JObject
                {
                    ["world"] = new JObject
                    {
                        ["x"] = transform.position.x,
                        ["y"] = transform.position.y,
                        ["z"] = transform.position.z
                    },
                    ["local"] = new JObject
                    {
                        ["x"] = transform.localPosition.x,
                        ["y"] = transform.localPosition.y,
                        ["z"] = transform.localPosition.z
                    }
                }
            };
        }
    }

    /// <summary>
    /// Tool for rotating a GameObject in the Unity Editor.
    /// Supports world/local space and absolute/relative rotation using Euler angles.
    /// </summary>
    public class RotateGameObjectTool : McpToolBase
    {
        public RotateGameObjectTool()
        {
            Name = "rotate_gameobject";
            Description = "Rotates a GameObject using Euler angles. Supports world/local space and absolute/relative rotation.";
            IsAsync = false;
        }

        public override JObject Execute(JObject parameters)
        {
            // Find the GameObject
            var findResult = TransformToolUtils.FindGameObject(parameters);
            if (findResult.Error != null)
                return findResult.Error;

            GameObject gameObject = findResult.GameObject;
            Transform transform = gameObject.transform;

            // Extract rotation (Euler angles)
            JObject rotationObj = parameters["rotation"] as JObject;
            if (rotationObj == null)
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    "Required parameter 'rotation' not provided",
                    "validation_error"
                );
            }

            Vector3 eulerAngles = new Vector3(
                rotationObj["x"]?.ToObject<float>() ?? 0f,
                rotationObj["y"]?.ToObject<float>() ?? 0f,
                rotationObj["z"]?.ToObject<float>() ?? 0f
            );

            // Get space and relative flags
            string space = parameters["space"]?.ToObject<string>() ?? "world";
            bool relative = parameters["relative"]?.ToObject<bool>() ?? false;

            // Record undo
            Undo.RecordObject(transform, "Rotate GameObject");

            // Apply the rotation
            if (relative)
            {
                // Relative rotation - add to current rotation
                Space unitySpace = space.ToLower() == "local" ? Space.Self : Space.World;
                transform.Rotate(eulerAngles, unitySpace);
            }
            else
            {
                // Absolute rotation - set directly
                if (space.ToLower() == "local")
                    transform.localEulerAngles = eulerAngles;
                else
                    transform.eulerAngles = eulerAngles;
            }

            EditorUtility.SetDirty(gameObject);

            // Return result with new rotation
            return new JObject
            {
                ["success"] = true,
                ["type"] = "text",
                ["message"] = $"GameObject '{gameObject.name}' rotated successfully.",
                ["instanceId"] = gameObject.GetInstanceID(),
                ["name"] = gameObject.name,
                ["path"] = TransformToolUtils.GetGameObjectPath(gameObject),
                ["rotation"] = new JObject
                {
                    ["world"] = new JObject
                    {
                        ["x"] = transform.eulerAngles.x,
                        ["y"] = transform.eulerAngles.y,
                        ["z"] = transform.eulerAngles.z
                    },
                    ["local"] = new JObject
                    {
                        ["x"] = transform.localEulerAngles.x,
                        ["y"] = transform.localEulerAngles.y,
                        ["z"] = transform.localEulerAngles.z
                    }
                }
            };
        }
    }

    /// <summary>
    /// Tool for scaling a GameObject in the Unity Editor.
    /// Supports absolute and relative (multiplicative) scaling.
    /// </summary>
    public class ScaleGameObjectTool : McpToolBase
    {
        public ScaleGameObjectTool()
        {
            Name = "scale_gameobject";
            Description = "Scales a GameObject. Supports absolute and relative (multiplicative) scaling.";
            IsAsync = false;
        }

        public override JObject Execute(JObject parameters)
        {
            // Find the GameObject
            var findResult = TransformToolUtils.FindGameObject(parameters);
            if (findResult.Error != null)
                return findResult.Error;

            GameObject gameObject = findResult.GameObject;
            Transform transform = gameObject.transform;

            // Extract scale
            JObject scaleObj = parameters["scale"] as JObject;
            if (scaleObj == null)
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    "Required parameter 'scale' not provided",
                    "validation_error"
                );
            }

            Vector3 scale = new Vector3(
                scaleObj["x"]?.ToObject<float>() ?? 1f,
                scaleObj["y"]?.ToObject<float>() ?? 1f,
                scaleObj["z"]?.ToObject<float>() ?? 1f
            );

            // Get relative flag
            bool relative = parameters["relative"]?.ToObject<bool>() ?? false;

            // Record undo
            Undo.RecordObject(transform, "Scale GameObject");

            // Apply the scale
            if (relative)
            {
                // Relative scale - multiply current scale
                transform.localScale = Vector3.Scale(transform.localScale, scale);
            }
            else
            {
                // Absolute scale - set directly
                transform.localScale = scale;
            }

            EditorUtility.SetDirty(gameObject);

            // Return result with new scale
            return new JObject
            {
                ["success"] = true,
                ["type"] = "text",
                ["message"] = $"GameObject '{gameObject.name}' scaled successfully.",
                ["instanceId"] = gameObject.GetInstanceID(),
                ["name"] = gameObject.name,
                ["path"] = TransformToolUtils.GetGameObjectPath(gameObject),
                ["scale"] = new JObject
                {
                    ["x"] = transform.localScale.x,
                    ["y"] = transform.localScale.y,
                    ["z"] = transform.localScale.z
                }
            };
        }
    }

    /// <summary>
    /// Tool for setting a GameObject's full transform (position, rotation, scale) in one operation.
    /// All parameters are optional - only provided values will be changed.
    /// </summary>
    public class SetTransformTool : McpToolBase
    {
        public SetTransformTool()
        {
            Name = "set_transform";
            Description = "Sets a GameObject's transform (position, rotation, scale) in one operation. All transform properties are optional.";
            IsAsync = false;
        }

        public override JObject Execute(JObject parameters)
        {
            // Find the GameObject
            var findResult = TransformToolUtils.FindGameObject(parameters);
            if (findResult.Error != null)
                return findResult.Error;

            GameObject gameObject = findResult.GameObject;
            Transform transform = gameObject.transform;

            // Check that at least one transform property is provided
            JObject positionObj = parameters["position"] as JObject;
            JObject rotationObj = parameters["rotation"] as JObject;
            JObject scaleObj = parameters["scale"] as JObject;

            if (positionObj == null && rotationObj == null && scaleObj == null)
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    "At least one of 'position', 'rotation', or 'scale' must be provided",
                    "validation_error"
                );
            }

            // Get space flag (applies to position and rotation)
            string space = parameters["space"]?.ToObject<string>() ?? "world";
            bool isLocal = space.ToLower() == "local";

            // Record undo
            Undo.RecordObject(transform, "Set Transform");

            // Apply position if provided
            if (positionObj != null)
            {
                Vector3 position = new Vector3(
                    positionObj["x"]?.ToObject<float>() ?? (isLocal ? transform.localPosition.x : transform.position.x),
                    positionObj["y"]?.ToObject<float>() ?? (isLocal ? transform.localPosition.y : transform.position.y),
                    positionObj["z"]?.ToObject<float>() ?? (isLocal ? transform.localPosition.z : transform.position.z)
                );

                if (isLocal)
                    transform.localPosition = position;
                else
                    transform.position = position;
            }

            // Apply rotation if provided
            if (rotationObj != null)
            {
                Vector3 eulerAngles = new Vector3(
                    rotationObj["x"]?.ToObject<float>() ?? (isLocal ? transform.localEulerAngles.x : transform.eulerAngles.x),
                    rotationObj["y"]?.ToObject<float>() ?? (isLocal ? transform.localEulerAngles.y : transform.eulerAngles.y),
                    rotationObj["z"]?.ToObject<float>() ?? (isLocal ? transform.localEulerAngles.z : transform.eulerAngles.z)
                );

                if (isLocal)
                    transform.localEulerAngles = eulerAngles;
                else
                    transform.eulerAngles = eulerAngles;
            }

            // Apply scale if provided
            if (scaleObj != null)
            {
                Vector3 scale = new Vector3(
                    scaleObj["x"]?.ToObject<float>() ?? transform.localScale.x,
                    scaleObj["y"]?.ToObject<float>() ?? transform.localScale.y,
                    scaleObj["z"]?.ToObject<float>() ?? transform.localScale.z
                );
                transform.localScale = scale;
            }

            EditorUtility.SetDirty(gameObject);

            // Return result with full transform data
            return new JObject
            {
                ["success"] = true,
                ["type"] = "text",
                ["message"] = $"GameObject '{gameObject.name}' transform updated successfully.",
                ["instanceId"] = gameObject.GetInstanceID(),
                ["name"] = gameObject.name,
                ["path"] = TransformToolUtils.GetGameObjectPath(gameObject),
                ["transform"] = new JObject
                {
                    ["position"] = new JObject
                    {
                        ["world"] = new JObject
                        {
                            ["x"] = transform.position.x,
                            ["y"] = transform.position.y,
                            ["z"] = transform.position.z
                        },
                        ["local"] = new JObject
                        {
                            ["x"] = transform.localPosition.x,
                            ["y"] = transform.localPosition.y,
                            ["z"] = transform.localPosition.z
                        }
                    },
                    ["rotation"] = new JObject
                    {
                        ["world"] = new JObject
                        {
                            ["x"] = transform.eulerAngles.x,
                            ["y"] = transform.eulerAngles.y,
                            ["z"] = transform.eulerAngles.z
                        },
                        ["local"] = new JObject
                        {
                            ["x"] = transform.localEulerAngles.x,
                            ["y"] = transform.localEulerAngles.y,
                            ["z"] = transform.localEulerAngles.z
                        }
                    },
                    ["scale"] = new JObject
                    {
                        ["x"] = transform.localScale.x,
                        ["y"] = transform.localScale.y,
                        ["z"] = transform.localScale.z
                    }
                }
            };
        }
    }

    /// <summary>
    /// Utility class for common transform tool operations
    /// </summary>
    internal static class TransformToolUtils
    {
        /// <summary>
        /// Result of finding a GameObject
        /// </summary>
        public struct FindResult
        {
            public GameObject GameObject;
            public JObject Error;
        }

        /// <summary>
        /// Find a GameObject by instanceId or objectPath from parameters
        /// </summary>
        public static FindResult FindGameObject(JObject parameters)
        {
            int? instanceId = parameters["instanceId"]?.ToObject<int?>();
            string objectPath = parameters["objectPath"]?.ToObject<string>();

            GameObject gameObject = null;
            string identifierInfo = "";

            if (instanceId.HasValue)
            {
                gameObject = EditorUtility.InstanceIDToObject(instanceId.Value) as GameObject;
                identifierInfo = $"instance ID {instanceId.Value}";
            }
            else if (!string.IsNullOrEmpty(objectPath))
            {
                gameObject = GameObject.Find(objectPath);
                identifierInfo = $"path '{objectPath}'";
            }
            else
            {
                return new FindResult
                {
                    Error = McpUnitySocketHandler.CreateErrorResponse(
                        "Either 'instanceId' or 'objectPath' must be provided",
                        "validation_error"
                    )
                };
            }

            if (gameObject == null)
            {
                return new FindResult
                {
                    Error = McpUnitySocketHandler.CreateErrorResponse(
                        $"GameObject not found with {identifierInfo}",
                        "not_found_error"
                    )
                };
            }

            return new FindResult { GameObject = gameObject };
        }

        /// <summary>
        /// Get the hierarchy path of a GameObject
        /// </summary>
        public static string GetGameObjectPath(GameObject obj)
        {
            if (obj == null) return null;
            string path = "/" + obj.name;
            while (obj.transform.parent != null)
            {
                obj = obj.transform.parent.gameObject;
                path = "/" + obj.name + path;
            }
            return path;
        }
    }
}

File: Workpiece/Summoner/LocalPackages/com.gamelovers.mcp-unity/Server~/src/tools/transformTools.ts
import * as z from 'zod';
import { McpUnity } from '../unity/mcpUnity.js';
import { Logger } from '../utils/logger.js';
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { McpUnityError, ErrorType } from '../utils/errors.js';
import { CallToolResult } from '@modelcontextprotocol/sdk/types.js';

// Build a fresh Vector3 schema per field to avoid local JSON pointer refs
// like "#/properties/position" in generated JSON schema.
function createVector3Schema() {
  return z.object({
    x: z.number().describe('X component'),
    y: z.number().describe('Y component'),
    z: z.number().describe('Z component')
  });
}

// ============================================================================
// move_gameobject Tool
// ============================================================================

const moveToolName = 'move_gameobject';
const moveToolDescription = 'Moves a GameObject to a new position. Supports world/local space and absolute/relative positioning.';
const moveParamsSchema = z.object({
  instanceId: z.number().optional().describe('The instance ID of the GameObject to move'),
  objectPath: z.string().optional().describe('The path of the GameObject in the hierarchy (alternative to instanceId)'),
  position: createVector3Schema().describe('The target position'),
  space: z.enum(['world', 'local']).default('world').describe('Coordinate space: "world" or "local"'),
  relative: z.boolean().default(false).describe('If true, adds to current position instead of setting absolute position')
});

/**
 * Registers the move_gameobject tool with the MCP server
 */
export function registerMoveGameObjectTool(server: McpServer, mcpUnity: McpUnity, logger: Logger) {
  logger.info(`Registering tool: ${moveToolName}`);

  server.tool(
    moveToolName,
    moveToolDescription,
    moveParamsSchema.shape,
    async (params: z.infer<typeof moveParamsSchema>) => {
      try {
        logger.info(`Executing tool: ${moveToolName}`, params);
        const result = await moveToolHandler(mcpUnity, params);
        logger.info(`Tool execution successful: ${moveToolName}`);
        return result;
      } catch (error) {
        logger.error(`Tool execution failed: ${moveToolName}`, error);
        throw error;
      }
    }
  );
}

async function moveToolHandler(mcpUnity: McpUnity, params: z.infer<typeof moveParamsSchema>): Promise<CallToolResult> {
  validateGameObjectIdentifier(params);

  const response = await mcpUnity.sendRequest({
    method: moveToolName,
    params: {
      instanceId: params.instanceId,
      objectPath: params.objectPath,
      position: params.position,
      space: params.space,
      relative: params.relative
    }
  });

  if (!response.success) {
    throw new McpUnityError(
      ErrorType.TOOL_EXECUTION,
      response.message || 'Failed to move GameObject'
    );
  }

  return {
    content: [{
      type: response.type,
      text: response.message || 'GameObject moved successfully'
    }]
  };
}

// ============================================================================
// rotate_gameobject Tool
// ============================================================================

const rotateToolName = 'rotate_gameobject';
const rotateToolDescription = 'Rotates a GameObject using Euler angles. Supports world/local space and absolute/relative rotation.';
const rotateParamsSchema = z.object({
  instanceId: z.number().optional().describe('The instance ID of the GameObject to rotate'),
  objectPath: z.string().optional().describe('The path of the GameObject in the hierarchy (alternative to instanceId)'),
  rotation: createVector3Schema().describe('The rotation in Euler angles (degrees)'),
  space: z.enum(['world', 'local']).default('world').describe('Coordinate space: "world" or "local"'),
  relative: z.boolean().default(false).describe('If true, adds to current rotation instead of setting absolute rotation')
});

/**
 * Registers the rotate_gameobject tool with the MCP server
 */
export function registerRotateGameObjectTool(server: McpServer, mcpUnity: McpUnity, logger: Logger) {
  logger.info(`Registering tool: ${rotateToolName}`);

  server.tool(
    rotateToolName,
    rotateToolDescription,
    rotateParamsSchema.shape,
    async (params: z.infer<typeof rotateParamsSchema>) => {
      try {
        logger.info(`Executing tool: ${rotateToolName}`, params);
        const result = await rotateToolHandler(mcpUnity, params);
        logger.info(`Tool execution successful: ${rotateToolName}`);
        return result;
      } catch (error) {
        logger.error(`Tool execution failed: ${rotateToolName}`, error);
        throw error;
      }
    }
  );
}

async function rotateToolHandler(mcpUnity: McpUnity, params: z.infer<typeof rotateParamsSchema>): Promise<CallToolResult> {
  validateGameObjectIdentifier(params);

  const response = await mcpUnity.sendRequest({
    method: rotateToolName,
    params: {
      instanceId: params.instanceId,
      objectPath: params.objectPath,
      rotation: params.rotation,
      space: params.space,
      relative: params.relative
    }
  });

  if (!response.success) {
    throw new McpUnityError(
      ErrorType.TOOL_EXECUTION,
      response.message || 'Failed to rotate GameObject'
    );
  }

  return {
    content: [{
      type: response.type,
      text: response.message || 'GameObject rotated successfully'
    }]
  };
}

// ============================================================================
// scale_gameobject Tool
// ============================================================================

const scaleToolName = 'scale_gameobject';
const scaleToolDescription = 'Scales a GameObject. Supports absolute and relative (multiplicative) scaling.';
const scaleParamsSchema = z.object({
  instanceId: z.number().optional().describe('The instance ID of the GameObject to scale'),
  objectPath: z.string().optional().describe('The path of the GameObject in the hierarchy (alternative to instanceId)'),
  scale: createVector3Schema().describe('The scale values'),
  relative: z.boolean().default(false).describe('If true, multiplies current scale instead of setting absolute scale')
});

/**
 * Registers the scale_gameobject tool with the MCP server
 */
export function registerScaleGameObjectTool(server: McpServer, mcpUnity: McpUnity, logger: Logger) {
  logger.info(`Registering tool: ${scaleToolName}`);

  server.tool(
    scaleToolName,
    scaleToolDescription,
    scaleParamsSchema.shape,
    async (params: z.infer<typeof scaleParamsSchema>) => {
      try {
        logger.info(`Executing tool: ${scaleToolName}`, params);
        const result = await scaleToolHandler(mcpUnity, params);
        logger.info(`Tool execution successful: ${scaleToolName}`);
        return result;
      } catch (error) {
        logger.error(`Tool execution failed: ${scaleToolName}`, error);
        throw error;
      }
    }
  );
}

async function scaleToolHandler(mcpUnity: McpUnity, params: z.infer<typeof scaleParamsSchema>): Promise<CallToolResult> {
  validateGameObjectIdentifier(params);

  const response = await mcpUnity.sendRequest({
    method: scaleToolName,
    params: {
      instanceId: params.instanceId,
      objectPath: params.objectPath,
      scale: params.scale,
      relative: params.relative
    }
  });

  if (!response.success) {
    throw new McpUnityError(
      ErrorType.TOOL_EXECUTION,
      response.message || 'Failed to scale GameObject'
    );
  }

  return {
    content: [{
      type: response.type,
      text: response.message || 'GameObject scaled successfully'
    }]
  };
}

// ============================================================================
// set_transform Tool
// ============================================================================

const setTransformToolName = 'set_transform';
const setTransformToolDescription = 'Sets a GameObject\'s transform (position, rotation, scale) in one operation. All transform properties are optional.';
function createSetTransformParamsShape() {
  return {
    instanceId: z.number().optional().describe('The instance ID of the GameObject'),
    objectPath: z.string().optional().describe('The path of the GameObject in the hierarchy (alternative to instanceId)'),
    position: createVector3Schema().optional().describe('The position to set'),
    rotation: createVector3Schema().optional().describe('The rotation in Euler angles (degrees)'),
    scale: createVector3Schema().optional().describe('The scale to set'),
    space: z.enum(['world', 'local']).default('world').describe('Coordinate space for position and rotation: "world" or "local"')
  };
}

const setTransformParamsSchema = z.object({
  ...createSetTransformParamsShape()
}).refine(
  data => data.position !== undefined || data.rotation !== undefined || data.scale !== undefined,
  { message: 'At least one of position, rotation, or scale must be provided' }
);

/**
 * Registers the set_transform tool with the MCP server
 */
export function registerSetTransformTool(server: McpServer, mcpUnity: McpUnity, logger: Logger) {
  logger.info(`Registering tool: ${setTransformToolName}`);

  server.tool(
    setTransformToolName,
    setTransformToolDescription,
    // Use base shape without refine for MCP schema registration
    createSetTransformParamsShape(),
    async (params: any) => {
      try {
        logger.info(`Executing tool: ${setTransformToolName}`, params);
        const result = await setTransformToolHandler(mcpUnity, params);
        logger.info(`Tool execution successful: ${setTransformToolName}`);
        return result;
      } catch (error) {
        logger.error(`Tool execution failed: ${setTransformToolName}`, error);
        throw error;
      }
    }
  );
}

async function setTransformToolHandler(mcpUnity: McpUnity, params: any): Promise<CallToolResult> {
  validateGameObjectIdentifier(params);

  // Validate that at least one transform property is provided
  if (!params.position && !params.rotation && !params.scale) {
    throw new McpUnityError(
      ErrorType.VALIDATION,
      'At least one of position, rotation, or scale must be provided'
    );
  }

  const response = await mcpUnity.sendRequest({
    method: setTransformToolName,
    params: {
      instanceId: params.instanceId,
      objectPath: params.objectPath,
      position: params.position,
      rotation: params.rotation,
      scale: params.scale,
      space: params.space
    }
  });

  if (!response.success) {
    throw new McpUnityError(
      ErrorType.TOOL_EXECUTION,
      response.message || 'Failed to set transform'
    );
  }

  return {
    content: [{
      type: response.type,
      text: response.message || 'Transform updated successfully'
    }]
  };
}

// ============================================================================
// Helper Functions
// ============================================================================

/**
 * Validates that either instanceId or objectPath is provided
 */
function validateGameObjectIdentifier(params: { instanceId?: number; objectPath?: string }) {
  if ((params.instanceId === undefined || params.instanceId === null) &&
      (!params.objectPath || params.objectPath.trim() === '')) {
    throw new McpUnityError(
      ErrorType.VALIDATION,
      "Either 'instanceId' or 'objectPath' must be provided"
    );
  }
}

/**
 * Registers all transform tools with the MCP server
 */
export function registerTransformTools(server: McpServer, mcpUnity: McpUnity, logger: Logger) {
  registerMoveGameObjectTool(server, mcpUnity, logger);
  registerRotateGameObjectTool(server, mcpUnity, logger);
  registerScaleGameObjectTool(server, mcpUnity, logger);
  registerSetTransformTool(server, mcpUnity, logger);
}
